File: Java/src/main/java/QuestionBank/Solution301_400/Solution345.java
package QuestionBank.Solution301_400;

/**
 * @author windows
 */
public class Solution345 {
    public String reverseVowels(String s) {
        int left = 0;
        int right = s.length() - 1;
        char[] chars = s.toCharArray();
        while (left < right) {
            while(left < right && Character.toLowerCase(chars[left]) != 'a' &&
                                  Character.toLowerCase(chars[left]) != 'e' &&
                                  Character.toLowerCase(chars[left]) != 'i' &&
                                  Character.toLowerCase(chars[left]) != 'o' &&
                                  Character.toLowerCase(chars[left]) != 'u'){
                left++;
            }
            while(left < right && Character.toLowerCase(chars[right]) != 'a' &&
                  Character.toLowerCase(chars[right]) != 'e' &&
                  Character.toLowerCase(chars[right]) != 'i' &&
                  Character.toLowerCase(chars[right]) != 'o' &&
                  Character.toLowerCase(chars[right]) != 'u'){
                right--;
            }
            if (left < right){
                char temp = chars[left];
                chars[left] = chars[right];
                chars[right] = temp;
                left++;
                right--;
            }
        }
        return new String(chars);
    }


    public static void main(String[] args) {
        Solution345 solution345 = new Solution345();
        System.out.println(solution345.reverseVowels("aA"));
    }
}

File: C++/src/Solution301_400/Solution345.cpp
//
// Created by windows on 2023/3/2.
//

#include "Solution345.h"
#include <functional>
#include <iostream>
#include <vector>
#include <unordered_set>
#include <unordered_map>
#include <string>
using namespace std;
#include <unordered_set>

string Solution345::reverseVowels(string s) {
    unordered_set<char> set{'a', 'e', 'i', 'o', 'u'};

    int left = 0, right = s.length() - 1;
    while (left < right) {
        while(left < right && !set.count(tolower(s[left]))) left++;
        while(left < right && !set.count(tolower(s[right]))) right--;
        char c = s[left];
        s[left] = s[right];
        s[right] = c;
        left++;
        right--;
    }
    return s;
}
